File: pintia/gplt_practice/l2_001.cpp
#include <cstdio>
#include <cstring>
#include <utility>
#include <queue>
#include <vector>
#include <algorithm>

using namespace std;

typedef pair<int, int> PII;

const int N = 510, M = N * N * 2, INF = 0x3f3f3f3f;
int n, m, s, d;
int w1[N];
int h[N], w2[M], e[M], ne[M], idx;
int dists[N], cnts[N], nums[N];
bool vis[N];
int prevs[N];

void add(int a, int b, int c){
    e[idx] = b, w2[idx] = c, ne[idx] = h[a], h[a] = idx++;
}

void dij(int src, int dest) {
    memset(dists, 0x3f, sizeof(dists));
    memset(prevs, -1, sizeof(prevs));
    dists[src] = 0;
    cnts[src] = 1;
    nums[src] = w1[src];
    priority_queue<PII, vector<PII>, greater<PII>> pq;
    pq.push({0, src});
    while (!pq.empty()) {
        PII u = pq.top(); pq.pop();
        int mn = u.second;
        if (vis[mn]) continue;
        vis[mn] = true;
        int d = u.first;
        for (int i = h[mn]; ~i; i = ne[i]) {
            int v = e[i];
            if (d + w2[i] < dists[v]) {
                dists[v] = d + w2[i];
                cnts[v] = cnts[mn];
                nums[v] = nums[mn] + w1[v];
                prevs[v] = mn;
                pq.push({dists[v], v});
            } else if (d + w2[i] == dists[v]) {
                if (nums[mn] + w1[v] > nums[v]) {
                    nums[v] = nums[mn] + w1[v];
                    prevs[v] = mn;
                }
                cnts[v] += cnts[mn];
            }
        }
    }
}

int main() {
    scanf("%d%d%d%d", &n, &m, &s, &d);
    for (int i = 0; i < n; ++i) scanf("%d", &w1[i]);
    memset(h, -1, sizeof(h));
    while (m--) {
        int a, b, c;
        scanf("%d%d%d", &a, &b, &c);
        add(a, b, c), add(b, a, c);
    }
    dij(s, d);
    printf("%d %d\n", cnts[d], nums[d]);
    vector<int> path;
    while (d != s) {
        path.push_back(d);
        d = prevs[d];
    }
    path.push_back(s);
    for (int i = path.size() - 1; i >= 0; --i) {
        printf("%d", path[i]);
        printf(i ? " " : "\n");
    }

    return 0;
}

File: pintia/gplt_practice/L2_001.java
import java.util.*;
import java.io.*;

public class L2_001 {
    static final int INF = 0x3f3f3f3f;
    static int n, m, s, d;
    static int[] w1;
    static int[] h, w2, e, ne;
    static int idx;
    static int[] dists, cnts, nums;
    static boolean[] vis;
    static int[] prevs;

    static void add(int a, int b, int c) {
        e[idx] = b; w2[idx] = c; ne[idx] = h[a]; h[a] = idx++;
    }

    static void dij(int src, int dest) {
        dists = new int[n];
        cnts = new int[n];
        nums = new int[n];
        vis = new boolean[n];
        prevs = new int[n];
        Arrays.fill(dists, INF);
        Arrays.fill(prevs, -1);
        dists[src] = 0;
        cnts[src] = 1;
        nums[src] = w1[src];
        Queue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.add(new int[]{0, src});
        while (!pq.isEmpty()) {
            int[] u = pq.poll();
            int mn = u[1];
            if (vis[mn]) continue;
            vis[mn] = true;
            int d = u[0];
            for (int i = h[mn]; i != -1; i = ne[i]) {
                int v = e[i];
                if (d + w2[i] < dists[v]) {
                    dists[v] = d + w2[i];
                    cnts[v] = cnts[mn];
                    nums[v] = nums[mn] + w1[v];
                    prevs[v] = mn;
                    pq.add(new int[]{dists[v], v});
                } else if (d + w2[i] == dists[v]) {
                    if (nums[mn] + w1[v] > nums[v]) {
                        nums[v] = nums[mn] + w1[v];
                        prevs[v] = mn;
                    }
                    cnts[v] += cnts[mn];
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        in.nextToken(); n = (int) in.nval;
        in.nextToken(); m = (int) in.nval;
        in.nextToken(); s = (int) in.nval;
        in.nextToken(); d = (int) in.nval;
        PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out));
        h = new int[n];
        Arrays.fill(h, -1);
        w1 = new int[n];
        w2 = new int[m * 2];
        e = new int[m * 2];
        ne = new int[m * 2];
        for (int i = 0; i < n; ++i) {
            in.nextToken(); w1[i] = (int) in.nval;
        }
        while (m-- > 0) {
            in.nextToken(); int a = (int) in.nval;
            in.nextToken(); int b = (int) in.nval;
            in.nextToken(); int c = (int) in.nval;
            add(a, b, c); add(b, a, c);
        }
        dij(s, d);
        out.println(cnts[d] + " " + nums[d]);
        StringBuilder path = new StringBuilder();
        while (d != s) {
            path.append(d).append(' ');
            d = prevs[d];
        }
        path.append(s);
        out.println(path.reverse());
        out.flush();
        out.close();
    }
}
